File: balanced_binary_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        if root == None: return True
        node = root
        while node != None:
            if abs( self.depth(node.left) - self.depth(node.right) ) > 1:
                return False
            else:
                return self.isBalanced(node.left) and self.isBalanced(node.right)
    
    def depth(self, root):
        if root == None: return 0
        else:
            return 1 + max( self.depth(root.left), self.depth(root.right) )

File: cleanCode50/BalancedBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        return maxDepth(root) != -1;
    }

    private int maxDepth(TreeNode root) {
    	if (root == null) {
    		return 0;
    	}
    	int leftDepth = maxDepth(root.left);
    	if (leftDepth == -1) {
    		return -1;
    	}
    	int rightDepth = maxDepth(root.right);
    	if (rightDepth == -1) {
    		return -1;
    	}
    	if (Math.abs(leftDepth - rightDepth) <= 1) {
    		return 1 + Math.max(leftDepth, rightDepth);
    	} else {
    		return -1;
    	}
    }
}